src/main/java/mate/academy/service/impl/BookServiceImpl.java

```diff
@@ -57,14 +66,26 @@ public void deleteById(Long id) {
 
     @Override
     public List<BookDto> findByParameters(BookSearchParametersDto bookSearchDto,
-                                          Pageable pageable) {
+                                                            Pageable pageable) {
         return bookRepository.findAll(builder.buildSpecification(bookSearchDto), pageable)
                 .stream()
                 .map(bookMapper::toDto)
                 .toList();
     }
 
+    @Override
+    public List<BookDtoWithoutCategoryIds> findAllByCategoryId(Long categoryId) {
+        return bookRepository.findByCategories_Id(categoryId).stream()
+                .map(bookMapper::toDtoWithoutCategories)
+                .toList();
+    }
+
     private String getNotFoundMessage(Long id) {
         return String.format("Book with id: %s not found", id);
     }
+
+    private List<Category> categoriesIdToCategories(List<Long> categories) {
+        return categoryRepository.findAllById(Optional.ofNullable(categories)
```

Comment:
you don't need to load complete categories

instead you can leverage repository `getReference` method in order to create mapping

Reply:
changed (if I got it correctly).